tools/bbbsupervisor/README.md

```diff
@@ -68,3 +37,58 @@ RestartSec=10
+
+- a `logWatcher` watching systemd logs for a specific service (e.g. `bitcoind.service`) via `journalctl`
+- a `prometheusWatcher` watching a specific measurement exposed via the Prometheus API
+
+### The `logWatcher`
+
+For each systemd service, a `logWatcher` is started in its own goroutine. It starts to `--follow` the systemd log of that unit via `journalctl`. `stdout` output is written to an `eventWriter` which parses a line (sometimes also multiple lines, known issue) into an event by performing string matching on it. Each `watcherEvent` gets assigned a trigger. The event is passed into a _event channel_ called `events`. `stderr` output is written to an `errWritter` which passes all line(s) read into an _error channel_ called `errs`.
+
+### The `prometheusWatcher`
+
+For each Prometheus value to watch a `prometheusWatcher` is started in its own goroutine. The `prometheusWatcher` queries a specific `measure` or `expression`. It passes a watcherEvent into the `events` channel with the `measure` and the measured `value`. The watcher then sleeps and queries again after waking back up. The query `interval` can be set for each `prometheusWatcher`.
+
+### Event handling
+
+Events are indefinitely read from the channels (`errs`, `events`) in the `eventLoop()` function. First errors from the `errs` channel are read (if existent) and a _panic_ is thrown (currently not _recovered_ yet). Then `events` is read and the triggers are handled in the respective handle functions. Then the event handling loop restarts.
+
+### Currently handled event triggers (business logic)
+
+There are currently three triggers handled: `triggerElectrsFullySynced`, `triggerElectrsNoBitcoindConnectivity` and `triggerPrometheusBitcoindIDB`. I propose to document and propose triggers (including action and rationale) in a table.
+
+| trigger | fired when | action performed | rationale |
+| ---  | --- | --- | --- |
+| `triggerElectrsFullySynced` (logWatcher) | when electrs is fully synced. electrs stdout contains `"finished full compaction"`. | Restart electrs. (this is not tested, just taken from previous versions) | Maybe Stadicus knows? |
+| `triggerElectrsNoBitcoindConnectivity` (logWatcher) | when the `.cookie.env` file changes without electrs noticing. electrs stdout `"WARN - reconnecting to bitcoind: no reply from daemon"` | restart electrs which, via script (7e7a09c5907a68783971a03e52d866b738f75462) recreates the cookie file| electrs sometime gets out of sync with the bitcoind .cookie file |
+| `triggerPrometheusBitcoindIDB` (prometheusWatcher) | every time a Prometheus measurement for `"bitcoind_ibd"` has been made | initial trigger: run `bbbconfig.sh set bitcoin_idb true/false`. not changed: nothing, changed: run `bbbconfig.sh set bitcoin_idb true/false` | Setting e.g. a different db_cache and stop lightningd and electrs on initial block download |
+
```

Comment:
Nit: You might want to add the newly introduced trigger `triggerMiddlewareNoBitcoindConnectivity` here as well to keep docs consistent.

Reply:
Rephrased the whole table a bit (it was too big) and included the `triggerMiddlewareNoBitcoindConnectivity` trigger.

Reply:
Way better now!